### ISSUE TITLE
add option for silent boot

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#32555]
## Issue description

I don't want to see the boot messages on my desktop systems. The Login manager should appear directly after the Grub menu (when there are no critical errors).

I tried several things like:

```
  # less verbose boot log
  boot.consoleLogLevel = 3;
  boot.kernelParams = [ "quiet" "udev.log_priority=3" ];
```
and
```
  boot.loader.grub.extraConfig = "GRUB_CMDLINE_LINUX_DEFAULT=\"quiet loglevel=3 vga=current\"";
```
It has no effect.

Source: https://wiki.archlinux.org/index.php/silent_boot

### Steps to reproduce

Have NixOS system with Grub 2.

![virtualbox_nixos_11_12_2017_06_02_11](https://user-images.githubusercontent.com/91113/33816452-48cd89c8-de39-11e7-8511-dd0ef5eea055.png)

## Technical details

 - system: `"x86_64-linux"`
 - host os: `Linux 4.9.66, NixOS, 17.09.2281.b4a0c011e81 (Hummingbird)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 1.11.15`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`

**Comment by davidak** [comment]
Ubuntu uses **Plymouth**.

It can get enabled in NixOS but looks broken (that is another issue).

```
  boot.plymouth.enable = true;
```

Should we enable it by default for any `displayManager`?

**Comment by michaelpj** [comment]
It's also the case that the default Plymouth theme is incredibly ugly (I'm reluctant to make it the default while it is so). I've been meaning to submit a PR to make the default a nicer one, I should get around to that.

**Comment by davidak** [comment]
A NixOS theme would be nice.

There are also others that can get packaged.

Tux:
https://www.gnome-look.org/p/1189328/
https://www.gnome-look.org/p/1000026/

this would be nice for apple devices: https://www.gnome-look.org/p/1009320/

Windows boot screens are funny:
https://www.gnome-look.org/p/1202357/
https://www.gnome-look.org/p/1202116/

**Comment by michaelpj** [comment]
I [packaged](https://github.com/NixOS/nixpkgs/pull/31378) a nice one from KDE-land that works with NixOS branding, just need to make it available by default.

**Comment by gnidorah** [comment]
https://github.com/NixOS/nixpkgs/pull/32578

**Comment by michaelpj** [comment]
That still doesn't do anything about the stage 1 messages, though.

**Comment by gnidorah** [comment]
Yep. /cc @abbradar Also https://github.com/NixOS/nixpkgs/issues/32556

**Comment by sixpointzero** [comment]
It seems that only messages from systemd are hidden if `quiet` and `loglevel=4` are present in the kernel parameters with `quiet` mentioned before `loglevel=4` as mentioned [here](https://wiki.archlinux.org/index.php/Silent_boot#Kernel_parameters). However, I'm not sure whether the NixOS configuration option `boot.consoleLogLevel` ensures that `quiet` is present before `loglevel=4` is appended.
Moreover, as @michaelpj mentioned, I'm also not sure how I can hide the non-critical `NixOS Stage 1` and `NixOS Stage 2` messages.

**Comment by sixpointzero** [comment]
The description of the `boot.consoleLogLevel` option should also be changed from ` ... a priority numerically less than ...` to ` ... a priority numerically greater than ...`.

**Comment by peterhoeg** [comment]
> We probably need to patch those scripts to not use echo but a logging function where the verbosity can be controlled.

That's what I did locally but I really shoudl clean that up and upstream it.

**Comment by p4block** [comment]
I tried to set up quiet boot but still got the stage1 logging, thankfully I quickly encountered this post :+1: 

I found [the patch to the initrd sh script in your nixpkgs fork](
https://github.com/NixOS/nixpkgs/compare/master...peterhoeg:quiet_boot) @peterhoeg, thanks a lot. I'll be stealing that for the moment. 

Should be enough to get lazy fbcon init to work, which would allow keeping the grub splash image until getty or the DM start.

**Comment by jonas-schievink** [comment]
As mentioned in that Discourse thread, this is not yet fixed for the stage2 script, so reopening.

**Comment by mainrs** [comment]
taking a look at the stage-1 script: how does the `@verbose@` placeholder/flag work? I've never seen that syntax before.

**Comment by arbv** [comment]
not stale

**Comment by pyrotek45** [comment]
im surprised this isnt an open to have a silent boot up in ones nixos config

**Comment by smravec** [comment]
For anyone new to nixos wondering how to do this check out my dotfiles: https://github.com/smravec/nixos-config (look in the ./nixos/silent-boot/* and in ./nixos/configuration.nix files, and essentially what I have done is disabled the scripts that dont respect the kernel params and imported back a modified version of them that has all the lines with echo removed )

**Comment by ScottFreeCode** [comment]
This might be a silly question, but if I want those lines still being output/logged in case they're needed after the fact for troubleshooting or research or diagnosis or anything, and just want them not to be displayed at the time or (if this were possible) for the display of the animated boot screen, which already hides the later messages, to kick in sooner and hide even the earlier messages… Is that a reasonable goal, hiding the messages one way or another without actually removing them?

**Comment by smravec** [comment]
I dont think you can start plymouth before the stage 1 but you could instead of having them displayed redirect their output to some log file (which you could look at later in chroot if your system is not booting)

**Comment by samueldr** [comment]
Note that with EFI, modern linux will show the BGRT logo (e.g. manufacturer logo) until something needs to be printed to the fbcon.

So, one way to solve/workaround the problem here is to use `[ "console=tty2" ]`, or using `[ "fbcon=vc:2-6" "console=tty0" ]`.

I haven't re-verified the former (using tty2) since a long while, but that will allow a normal fbcon to exist there with a getty, so that may not be sufficient.

As for the latter solution (using fbcon=vc:2-6), this is removing fbcon from VT1, so in other word `ctrl+alt+f1` goes to a VT with nothing on it. So the BGRT will be present until something graphical takes over. My boots are entirely text-free. With the recently added grub changes, they are also flicker-free where possible (with `boot.loader.grub.timeoutStyle = false; `).

Briefly: two solutions:

 - Don't print to tty1/default VT's fbcon
 - Don't have an fbcon on VT1 (default VT)

**Comment by samueldr** [comment]
> if I want those lines still being output/logged in case they're needed after the fact for troubleshooting or research or diagnosis or anything

Good news, they are!

 - https://github.com/NixOS/nixpkgs/blob/2501addf2d61a3052128659883fee33435edd669/nixos/modules/system/boot/stage-1-init.sh#L163-L178

You should be able to see them with e.g. `journalctl -b` somewhere after `Run /init as init process`, since it's part of the `dmesg` output by being redirected to `kmsg`.

I believe that with the new systemd-based stage-1, it will also log to the journal, through a different mechanism [citation needed].
